torch/csrc/jit/passes/utils/subgraph_utils.cpp

```diff
@@ -9,6 +10,82 @@ bool hasSubgraph(Node* n) {
   return n->hasAttribute(attr::Subgraph);
 }
 
+std::vector<c10::optional<const Use>> gatherLastUses(
+    at::ArrayRef<Value*> values) {
+  return fmap(values, [&](Value* v) -> c10::optional<const Use> {
+    return firstOrLastUse(v, /*find_first*/ false);
+  });
+}
+
+// When merging a node into a subgraph, we wish to preserve all of the
+// aliasing properties of the node's outputs. It is difficult to track
+// the node or its contained nodes through all of the ir manipulation
+// involved in merging; it is pretty easy to uniquely identify the value
+// based on its uses. We can identify the value by its last use in the graph.
+// Values which do not have uses or which do not have a last use
+// outside of the subgraph to be merged into we do not need to track.
+struct ValueMapper {
+  ValueMapper(Node* to_merge, AliasDb& db, size_t subgraph_num_outputs) {
+    last_uses_ = gatherLastUses(to_merge->outputs());
+    subgraph_num_outputs_ = subgraph_num_outputs;
+    WithInsertPoint guard(to_merge);
+    auto g = to_merge->owningGraph();
+    // temporary node to put the aliasing properties of the node before its
+    // merged and destroyed
+    placeholder_node_ = g->insertNode(g->create(prim::Uninitialized, 0));
+    for (size_t i = 0; i < to_merge->outputs().size(); ++i) {
+      Value* existing = to_merge->outputs().at(i);
+      Value* new_value = placeholder_node_->insertOutput(i)->copyMetadata(
+          to_merge->outputs().at(i));
+      db.replaceWithNewValue(existing, new_value);
+    }
+  }
+
+  bool usesEqual(const Use& a, const Use& b) {
+    return a.user == b.user && a.offset == b.offset;
+  }
+
+  void copyAliasing(Node* merged_node, AliasDb& db) {
+    auto num_outputs = merged_node->outputs().size();
+    auto new_outputs = merged_node->outputs().slice(
+        subgraph_num_outputs_, num_outputs - subgraph_num_outputs_);
+    for (Value* v : new_outputs) {
+      auto maybe_last_use = firstOrLastUse(v, /*find_first*/ false);
+      // if it doesnt have a use it shouldnt have been added as output
+      TORCH_INTERNAL_ASSERT(maybe_last_use);
+      const Use last_use = *maybe_last_use;
+      size_t i = 0;
+      while (i < last_uses_.size() && last_uses_.at(i).has_value() &&
+             !usesEqual(*last_uses_.at(i), last_use)) {
+        ++i;
+      }
+      TORCH_INTERNAL_ASSERT(i != last_uses_.size());
+      db.replaceWithNewValue(placeholder_node_->outputs().at(i), v);
+    }
+    placeholder_node_->destroy();
+  }
+
+  std::vector<c10::optional<const Use>> last_uses_;
+  size_t subgraph_num_outputs_;
+  Node* placeholder_node_;
+};
+
+Node* executeSubgraphMergeAndUpdateAliasing(
```

Comment:
is it really worth it outlining `executeSubgraphMergeAndUpdateAliasing` ?

Reply:
it has two call sites

Reply:
yeah, my point was that it's a 3-line helper that requires callers to create lambdas that incidentally also capture the whole world (I can't easily tell what arguments those lambdas need). This makes the logic harder to follow. that's just my 2 cents.

Reply:
How would you refactor it ?

Reply:
I suppose I could just make it take in the two node ptrs and infer if we’re creating a new subgraph or merging

Reply:
I'd **consider** (not suggesting to actually do it) inlining it unless we think that `executeSubgraphMergeAndUpdateAliasing` might change frequently.

Reply:
Im gonna land to not re-incur wraath of CI but i might take a look at this later, as part of another PR/post release rush